### PR TITLE
Implement prioritization fees for program deploy using cli

### PR DESCRIPTION
**Pull request by godmodegalactus**
#### Problem

With cluster with high traffic where transactions are not processed without any prioritization fees. There is no way to set prioritization fee for program deployment.

#### Summary of Changes

Adding a way to set prioritization fees while doing program deployment using CLI.
Fixing a minor bug related to failed to join all the futures while program deployment.

Fixes #

### CODE REVIEW

**Comment by godmodegalactus** [comment]
No idea why clippy is failing, works fine on local. Could you give me access to solana kite.

File: cli/src/program.rs [hunk 1]
```diff
@@ -90,6 +91,7 @@ pub enum ProgramCliCommand {
         max_len: Option<usize>,
         allow_excessive_balance: bool,
         skip_fee_check: bool,
+        prioritization_fees: Option<u64>,
```
Comment (joncinque) [inline cli/src/program.rs:94]: nit: Can you call this `compute_unit_price`?

File: cli/src/program.rs [hunk 2]
```diff
@@ -99,6 +101,7 @@ pub enum ProgramCliCommand {
         sign_only: bool,
         dump_transaction_message: bool,
         blockhash_query: BlockhashQuery,
+        prioritization_fees: Option<u64>,
```
Comment (joncinque) [inline cli/src/program.rs:104]: same nit: Can you call this `compute_unit_price`?

File: cli/src/program.rs [hunk 3]
```diff
@@ -108,6 +111,7 @@ pub enum ProgramCliCommand {
         buffer_authority_signer_index: SignerIndex,
         max_len: Option<usize>,
         skip_fee_check: bool,
+        prioritization_fees: Option<u64>,
```
Comment (joncinque) [inline cli/src/program.rs:114]: same nit: Can you call this `compute_unit_price`?

File: cli/src/program.rs [hunk 4]
```diff
@@ -266,6 +281,17 @@ impl ProgramSubCommands for App<'_, '_> {
                                     "Upgrade authority [default: the default configured keypair]",
                                 ),
                         )
+                        .arg(
```
Comment (joncinque) [inline cli/src/program.rs:284]: Can you use https://github.com/anza-xyz/agave/blob/master/clap-utils/src/compute_unit_price.rs instead of redefining this everywhere?

File: cli/src/program.rs [hunk 5]
```diff
@@ -228,6 +232,17 @@ impl ProgramSubCommands for App<'_, '_> {
                                     [default: the length of the original deployed program]",
                                 ),
                         )
+                        .arg(
```
Comment (joncinque) [inline cli/src/program.rs:235]: Same here, please use https://github.com/anza-xyz/agave/blob/master/clap-utils/src/compute_unit_price.rs

File: cli/src/program.rs [hunk 6]
```diff
@@ -308,6 +334,17 @@ impl ProgramSubCommands for App<'_, '_> {
                                     "Maximum length of the upgradeable program \
                                     [default: the length of the original deployed program]",
                                 ),
+                        )
+                        .arg(
```
Comment (joncinque) [inline cli/src/program.rs:338]: Same here, please use https://github.com/anza-xyz/agave/blob/master/clap-utils/src/compute_unit_price.rs

File: cli/src/program.rs [hunk 7]
```diff
@@ -1329,22 +1381,29 @@ fn process_program_upgrade(
...
+    let upgrade_instruction = bpf_loader_upgradeable::upgrade(
+        &program_id,
+        &buffer_pubkey,
+        &upgrade_authority_signer.pubkey(),
+        &fee_payer_signer.pubkey(),
     );
 
+    let ixs = match prioritization_fees {
+        Some(fees) => vec![
+            compute_budget::ComputeBudgetInstruction::set_compute_unit_price(fees),
```
Comment (joncinque) [inline cli/src/program.rs:1399]: To avoid having people pay too much in priority fees, how about also setting the compute unit limit to the default for the bpf loader? https://github.com/anza-xyz/agave/blob/72d6d78d05738c3976c37de8eb250281898b7c1c/programs/bpf_loader/src/lib.rs#L60

File: cli/src/program.rs [hunk 8]
```diff
@@ -2209,11 +2274,14 @@ fn do_process_program_write_and_deploy(
     buffer_authority_signer: &dyn Signer,
     allow_excessive_balance: bool,
     skip_fee_check: bool,
+    prioritization_fees: Option<u64>,
 ) -> ProcessResult {
     let blockhash = rpc_client.get_latest_blockhash()?;
+    let prioritization_ix =
+        prioritization_fees.map(compute_budget::ComputeBudgetInstruction::set_compute_unit_price);
```
Comment (joncinque) [inline cli/src/program.rs:2281]: How about also setting the compute unit limit to avoid spending too much?

File: cli/src/program.rs [hunk 9]
```diff
@@ -2264,8 +2335,8 @@ fn do_process_program_write_and_deploy(
     };
 
     // Create and add write messages
-    let create_msg = |offset: u32, bytes: Vec<u8>| {
-        let instruction = if loader_id == &bpf_loader_upgradeable::id() {
+    let create_msg = |offset: u32, bytes: Vec<u8>, prioritization_ix: Option<Instruction>| {
```
Comment (joncinque) [inline cli/src/program.rs:2338]: nit: since this is a closure, you should be able to *not* require the new instruction param, but rather capture it from the outside, which should also simplify its usage

File: cli/src/program.rs [hunk 10]
```diff
@@ -2264,8 +2335,8 @@ fn do_process_program_write_and_deploy(
...
+        let instruction: Instruction = if loader_id == &bpf_loader_upgradeable::id() {
```
Comment (joncinque) [inline cli/src/program.rs:2339]: nit: is this change needed?

File: client/src/send_and_confirm_transactions_in_parallel.rs [hunk 11]
```diff
@@ -301,7 +301,16 @@ async fn sign_all_messages_and_send<T: Signers + ?Sized>(
...
+    match tokio::time::timeout(Duration::from_secs(120), join_all(futures)).await {
+        Ok(join_result) => {
+            join_result.into_iter().collect::<Result<_>>()?;
+        }
+        Err(_) => {
+            // It is fine if sending transactions to rpc is timed out here, we will retry it anyways,
+            // without this timeout cli waits indefinetly for all tasks to join
+            log::warn!("Timedout sending all transactions to RPC, will continue");
+        }
+    }
```
Comment (joncinque) [inline client/src/send_and_confirm_transactions_in_parallel.rs:313]: I might be missing something, but how is this change related?

File: cli/src/program.rs [hunk 12]
```diff
@@ -2364,75 +2444,84 @@ fn do_process_program_upgrade(
     buffer_pubkey: &Pubkey,
     buffer_signer: Option<&dyn Signer>,
     skip_fee_check: bool,
+    prioritization_fees: Option<u64>,
 ) -> ProcessResult {
     let blockhash = rpc_client.get_latest_blockhash()?;
+    let prioritization_ix =
+        prioritization_fees.map(compute_budget::ComputeBudgetInstruction::set_compute_unit_price);
```
Comment (joncinque) [inline cli/src/program.rs:2451]: Same here, let's include the compute unit limit instruction

File: cli/src/program.rs [hunk 13]
```diff
@@ -2364,75 +2444,84 @@ fn do_process_program_upgrade(
...
-    let (initial_message, write_messages, balance_needed) =
-        if let Some(buffer_signer) = buffer_signer {
-            // Check Buffer account to see if partial initialization has occurred
-            let (initial_instructions, balance_needed) = if let Some(account) = rpc_client
-                .get_account_with_commitment(&buffer_signer.pubkey(), config.commitment)?
-                .value
-            {
-                complete_partial_program_init(
-                    &bpf_loader_upgradeable::id(),
+    let (initial_message, write_messages, balance_needed) = if let Some(buffer_signer) =
```
Comment (joncinque) [inline cli/src/program.rs:2453]: Maybe I'm going cross-eyed this morning, but did anything actually change over here? It looks like only whitespace diffs

File: cli/src/program.rs [hunk 14]
```diff
@@ -2364,75 +2444,84 @@ fn do_process_program_upgrade(
...
-                    bytes,
-                );
-                Message::new_with_blockhash(
-                    &[instruction],
-                    Some(&fee_payer_signer.pubkey()),
-                    &blockhash,
-                )
+        let buffer_signer_pubkey = buffer_signer.pubkey();
+        let upgrade_authority_pubkey = upgrade_authority.pubkey();
+        let create_msg = |offset: u32, bytes: Vec<u8>, prioritization_ix: Option<Instruction>| {
```
Comment (joncinque) [inline cli/src/program.rs:2495]: Same here, let's remove the `Option<Instruction>` and just capture from the outside

File: cli/src/program.rs [hunk 15]
```diff
@@ -2364,75 +2444,84 @@ fn do_process_program_upgrade(
...
-        } else {
-            (None, vec![], 0)
-        };
+        (initial_message, write_messages, balance_needed)
+    } else {
+        (None, vec![], 0)
+    };
 
     // Create and add final message
     let final_message = Message::new_with_blockhash(
```
Comment (joncinque) [inline cli/src/program.rs:2527]: Can you also include the priority fees on this transaction?

File: cli/src/program_v4.rs [hunk 16]
```diff
@@ -585,16 +585,16 @@ pub fn process_deploy_program(
         };
 
     // Create and add write messages
-    let create_msg = |offset: u32, bytes: Vec<u8>| {
+    let create_msg = |offset: u32, bytes: Vec<u8>, _prioritization_fees_ix: Option<Instruction>| {
```
Comment (joncinque) [inline cli/src/program_v4.rs:588]: Once `create_msg` doesn't take the `Option<Instruction>`, you'll be able to revert these changes

**Review by joncinque — COMMENTED** [review]
Looks really good overall, thanks for your contribution! Mostly small things to fixup

**Comment by joncinque** [comment]
The clippy error is unrelated to your change, don't worry about that.

**Comment by joncinque** [comment]
CI seems to be picking up some old errors, do you need to rebase?

**Comment by Nagaprasadvr** [comment]
this is a duplicate of #144 opened last week
heres the issue 
https://github.com/anza-xyz/agave/issues/21
 @joncinque @godmodegalactus

**Comment by godmodegalactus** [comment]
> this is a duplicate of #144 opened last week heres the issue #21 @joncinque @godmodegalactus

Have you already finished working on it ?

**Comment by godmodegalactus** [comment]
> CI seems to be picking up some old errors, do you need to rebase?

Yes had rebased it, will try it again.

**Comment by Nagaprasadvr** [comment]
> > this is a duplicate of #144 opened last week heres the issue #21 @joncinque @godmodegalactus
> 
> Have you already finished working on it ?

Waiting for review

File: cli/src/program.rs (hunk 7 above)
Comment (jstarry) [reply cli/src/program.rs:1399]: Note that compute budget instructions themselves also consume 150 compute units each.

**Comment by godmodegalactus** [comment]
> > > this is a duplicate of #144 opened last week heres the issue #21 @joncinque @godmodegalactus
> > 
> > 
> > Have you already finished working on it ?
> 
> Waiting for review

Then go ahead merging your PR, let this PR stay open for a while because people might use this branch to deploy their programs.

**Comment by jstarry** [comment]
Fixing here: https://github.com/anza-xyz/agave/pull/364
